### PR TITLE
[WIP] Allow skin styles for sidebar context menu

### PR DESCRIPTION
**Pull request by ronso0**
**Closed and split up:**
Skins styles: #2337 
C++ changes: #2338
______________________

IMO the right-click menus like in Tracks, sidebar and cover art are an annoying issue because the (often ugly) OS themes crashes into our designed Mixxx skins.
For Tracks and cover art a custom style can be applied by addressing `QMenu` and `QMenu QCheckBox`.

To make this work for the library sidebar context menu, too, I set the sidebar widget as parent for each library feature's context menus. For now, I did this for History and Computer feature.
Please check if this is the right approach, then I'll add this to the other features as well. (I need to clean up a bit anyway)

Also I removed a c++ stylesheet for the Tracks > Crates menu. This allows to override the OS styled checkboxes in the Crates menu, as well.

qmenu-styles_LateNight

### CODE REVIEW

File: src/library/browse/browsefeature.cpp [hunk 1]
```diff
@@ -33,6 +33,7 @@ BrowseFeature::BrowseFeature(QObject* parent,
           m_proxyModel(&m_browseModel),
           m_pTrackCollection(pTrackCollection),
           m_pLastRightClickedItem(NULL),
+          m_pSidebarWidget(NULL),
```
Comment (daschuer) [inline src/library/browse/browsefeature.cpp]: Use nullptr

File: src/library/browse/browsefeature.cpp [hunk 2]
```diff
@@ -252,6 +257,11 @@ void BrowseFeature::activateChild(const QModelIndex& index) {
 }
 
 void BrowseFeature::onRightClickChild(const QPoint& globalPos, QModelIndex index) {
+    // Create the right-click menu and parent it to the sidebar widget in
+    // order to make it stylable with skin stylesheet rather than ugly OS styling.
+    // Parent to default NULL if there's no sidebar widget.
+    m_pMenu = new QMenu(m_pSidebarWidget);
```
Comment (daschuer) [inline src/library/browse/browsefeature.cpp]: It looks like as if you create a new QMenu on every right click.
Doesn't this work?
QMenu menu(m_pSidebarWidget);

File: src/library/browse/browsefeature.cpp [hunk 3]
```diff
@@ -265,23 +275,27 @@ void BrowseFeature::onRightClickChild(const QPoint& globalPos, QModelIndex index
...
         onLazyChildExpandation(index);
         return;
     }
 
     foreach (const QString& str, m_quickLinkList) {
         if (str == path) {
-             return;
+            // if path is a QuickLink:
+            // show remove action
+            m_pMenu->addAction(m_pRemoveQuickLinkAction);
```
Comment (daschuer) [inline src/library/browse/browsefeature.cpp]: Is this an unrelated change?

File: src/library/library.cpp [hunk 4]
```diff
@@ -240,6 +241,8 @@ void Library::bindSidebarWidget(WLibrarySidebar* pSidebarWidget) {
     pSidebarWidget->slotSetFont(m_trackTableFont);
     connect(this, SIGNAL(setTrackTableFont(QFont)),
             pSidebarWidget, SLOT(slotSetFont(QFont)));
+    m_pSetlogFeature->bindSidebarWidget(pSidebarWidget);
```
Comment (daschuer) [inline src/library/library.cpp]: Instead if name ach feature individual, you can use a loop over m_features

**Review by daschuer — CHANGES_REQUESTED** [review]
Thank you for investigate this. 
The solution looks good and I have left some comments.

File: src/library/browse/browsefeature.cpp (hunk 2 above)
Comment (ronso0) [reply src/library/browse/browsefeature.cpp]: nope, `QMenu menu(m_pSidebarWidget);` doesn't work. idk why

Is there a problem with creating a new menu? the old one is overwritten so there shouldn't be any side effect like increased memory usage, right?

File: src/library/browse/browsefeature.cpp (hunk 1 above)
Comment (ronso0) [reply src/library/browse/browsefeature.cpp]: fixed.

File: res/skins/Tango/style.qss [hunk 5]
```diff
@@ -2083,11 +2083,56 @@ WTrackProperty#SamplerTitle_mini {
  ######  Misc  ##################################################
 ##############################################################*/
 
-QToolTip {
+QToolTip,
+QMenu {
```
Comment (Be-ing) [inline res/skins/Tango/style.qss]: Does this still work with #2238? Do you need to style CueMenu and ColorMenu in addition to QMenu?

Comment (ronso0) [reply res/skins/Tango/style.qss]: It works, see the screenshot in the PR description which I composed from the hotcue label branch and this branch.
There are a few minor adjustments necessary though.
The LateNight screenshot only shows that it works in general, I'll take care of the other skins and the fine tuning when #2238 is merged

**Comment by Be-ing** [comment]
Some merge conflicts have developed.

Is this still a WIP?

**Comment by ronso0** [comment]
My next step is to aply the solution for Browse & History feature also to other features that have a sidebar context menu.
I'm not planning to overhaul the entire feature loading sequence here, like creating a separate method.

**Comment by ronso0** [comment]
> Is this still a WIP?

yes: I still need to implement the solution for AutoDJ, Playlists & Crates.

File: src/library/browse/browsefeature.cpp (hunk 2 above)
Comment (daschuer) [reply src/library/browse/browsefeature.cpp]: Only the plain pointer to the memory region is overwritten. The memory itself remains.  
You should use a smart pointer e.g. unique_ptr which automatically deletes the managed object if it is replaced see: https://de.cppreference.com/w/cpp/memory/unique_ptr/reset

Comment (ronso0) [reply src/library/browse/browsefeature.cpp]: Instead of using a unique pointer (which I don't really know how to handle, sry) I could also move `m_pMenu = new QMenu(m_pSidebarWidget);` to `void SetlogFeature::bindSidebarWidget` and clear/pouplate the menu in `void SetlogFeature::onRightClickChild`, right?

Comment (daschuer) [reply src/library/browse/browsefeature.cpp]: Sounds good. "new" is then called only once, right?

Comment (ronso0) [reply src/library/browse/browsefeature.cpp]: jup, once.
when `m_pSetlogFeature->bindSidebarWidget(pSidebarWidget);` is called from library.cpp when the feature is added.

**Comment by ronso0** [comment]
I guess I can revert 6d9cb58 "default m_pMenu to nullptr", right?
If there's no sidebar widget there'll b no right-click and no menu so it doesn't matter at all.

**Comment by ronso0** [comment]
The menu now looks nice for all library features, and also for the hotcue menu #2238 
Please test.

File: src/library/autodj/autodjfeature.cpp [hunk 6]
```diff
@@ -293,32 +306,31 @@ void AutoDJFeature::constructCrateChildModel() {
...
             connect(pAction.get(),
                     &QAction::triggered,
                     [=](bool) { m_crateMapper.map(); });
             crateMenu.addAction(pAction.get());
             pAction.release();
         }
-        QMenu contextMenu;
-        contextMenu.addMenu(&crateMenu);
-        contextMenu.exec(globalPos);
+        m_pMenu->addMenu(m_pCrateMenu);
```
Comment (nopeppermint) [inline src/library/autodj/autodjfeature.cpp]: change ```m_pMenu``` to ```m_pCrateMenu```

Comment (nopeppermint) [reply src/library/autodj/autodjfeature.cpp]: and there are more build errors in ```library.c``` around line 132 "BrowseFeature"

Comment (ronso0) [reply src/library/autodj/autodjfeature.cpp]: for AutoDJ `crateMenu` leftover was the issue.
`m_pMenu` and `m_pCrateMenu` are correct

`m_pBrowseFeature` is also fixed now

Comment (ronso0) [reply src/library/autodj/autodjfeature.cpp]: thx!

**Comment by nopeppermint** [comment]
nice

- styling for Shade could be better

**Comment by ronso0** [comment]
>     * styling for Shade could be better

Haa! probably because I didn't do anything in Shade so far..neither in Deere
@daschuer If yo don't have any style preferences I'll base the tooltip and context menu style on each scheme, like in LateNight.
@Be-ing What about Deere? Any ideas?

**Comment by ronso0** [comment]
Would be nice if someone with iTunes, Banshee &/ Rhythmbox could check the menu style in LateNight as I don't have any of those libraries

**Comment by daschuer** [comment]
I don't have any style preferences for the context menu of Shade. 

The playlist context menus of Traktor and Rhythmbox playlist are not styled.

File: src/library/setlogfeature.cpp [hunk 7]
```diff
@@ -71,6 +71,10 @@ void SetlogFeature::bindWidget(WLibrary* libraryWidget,
 
 void SetlogFeature::bindSidebarWidget(WLibrarySidebar* pSidebarWidget) {
     m_pSidebarWidget = pSidebarWidget;
```
Comment (Be-ing) [inline src/library/setlogfeature.cpp]: Add here:
```
if (m_pMenu != nullptr) {
    delete m_pMenu;
}
```
and likewise for the other features

File: src/library/library.cpp [hunk 8]
```diff
@@ -16,17 +16,7 @@
...
-#include "library/rhythmbox/rhythmboxfeature.h"
-#include "library/banshee/bansheefeature.h"
-#include "library/recording/recordingfeature.h"
-#include "library/itunes/itunesfeature.h"
-#include "library/mixxxlibraryfeature.h"
-#include "library/autodj/autodjfeature.h"
-#include "library/playlistfeature.h"
-#include "library/traktor/traktorfeature.h"
 #include "library/librarycontrol.h"
-#include "library/setlogfeature.h"
```
Comment (Be-ing) [inline src/library/library.cpp]: Why did you move these #includes?

**Review by Be-ing — CHANGES_REQUESTED** [review]
Nice work. The styling you have done in Tango and LateNight looks good to me. I don't have specific suggestions for Shade or Deere other than make the menus fit in to the rest of the skin.

File: src/library/setlogfeature.cpp (hunk 7 above)
Comment (Be-ing) [reply src/library/setlogfeature.cpp]: Actually I don't think you need to delete the menu. I think using QWidget::setParent would work:
```
if (m_pMenu == nullptr) {
   m_pMenu = new QMenu();
}
m_pMenu->setParent(pSidebarWidget);
```

File: src/library/library.cpp (hunk 8 above)
Comment (ronso0) [reply src/library/library.cpp]: Moving these includes to the header resolved build errors like "not able to resolve `AutoDJFeature* m_pAutoDJFeature;`" in library.h for example. And I thought it's more consistent to have all feature includes in one place.

Comment (Be-ing) [reply src/library/library.cpp]: Okay

File: src/library/setlogfeature.cpp [hunk 9]
```diff
@@ -12,13 +12,15 @@
...
 #include "widget/wlibrary.h"
+#include "widget/wlibrarysidebar.h"
 
 SetlogFeature::SetlogFeature(QObject* parent,
                              UserSettingsPointer pConfig,
                              TrackCollection* pTrackCollection)
         : BasePlaylistFeature(parent, pConfig, pTrackCollection, "SETLOGHOME"),
           m_playlistId(-1),
           m_libraryWidget(nullptr),
+          m_pSidebarWidget(NULL),
```
[the remark this answers is not in the document]
Comment (ronso0) [reply src/library/setlogfeature.cpp]: yeah will fix. leftover from the 1st commit I guess, I used nullprt everywhere else

File: src/library/setlogfeature.cpp (hunk 7 above)
Comment (ronso0) [reply src/library/setlogfeature.cpp]: for which case would this be useful?
now the menu is created only when there's a sidebar widegt passed from skinloader.
if the skin doesn't have a sidebar widget there's no sidebar bound to library.cpp > no sidebar passed on to library features, and there won't be any sidebar right-click either

Comment (Be-ing) [reply src/library/setlogfeature.cpp]: In case this `bindSidebarWidget` is ever called multiple times in the lifetime of the Feature. I don't think it is, but if that ever does happen in the future, the menu would have the wrong parent if this isn't changed.

Comment (ronso0) [reply src/library/setlogfeature.cpp]: hmkay. so the correct snippet for _all features_ would be
```
if (m_pMenu == nullptr) {
    m_pMenu = new QMenu(pSidebarWidget);
}
m_pMenu->setParent(pSidebarWidget);
```
because with this
```
if (m_pMenu == nullptr) {
    m_pMenu = new QMenu();
}
...
```
the menu wouldn't have a parent and there wouldn't be a way to style it, like before this PR

**Comment by ronso0** [comment]
> The playlist context menus of Traktor and Rhythmbox playlist are not styled.

okay, so I guess I'll need to add the menu changes to library/baseexternallibraryfeature.cpp/h

File: src/library/setlogfeature.cpp (hunk 7 above)
Comment (Be-ing) [reply src/library/setlogfeature.cpp]:
```
if (m_pMenu == nullptr) {
   m_pMenu = new QMenu();
}
m_pMenu->setParent(pSidebarWidget);
```
sets the parent regardless of whether m_pMenu existed previously. I suppose it wouldn't hurt to specify the parent to the QMenu constructor, but I am not sure it's necessary either.

Comment (ronso0) [reply src/library/setlogfeature.cpp]: forget my comment, I imagined an `else` at the start of the last line ;)

**Comment by ronso0** [comment]
> > The playlist context menus of Traktor and Rhythmbox playlist are not styled.
> 
> okay, so I guess I'll need to add the menu changes to library/baseexternallibraryfeature.cpp/h

I really don't have any idea how to accomplish this.
AFAICT BaseExternalLibraryFeature is loaded only once when each of the features is instantiated.
`addFeature` and `bindSidebarWidget` are separate functions in library.cpp.
How would I pass the sidebar widget pointer to each instance of BaseExternalLibraryFeature?

**Comment by Be-ing** [comment]
The menus in Deere do now show any indicator for submenus.

**Comment by ronso0** [comment]
> The menus in Deere do now show any indicator for submenus.

yeah I know.. this is still WIP. I stumble over many issues, also unrelated to this PR, especially with scaled Library widgets.

**Comment by ronso0** [comment]
Also this has grown beyond 'simply styling the menus'. I tried to match WEffectSelector and QToolTip and QMenu (in various places: hotcues, library header, tracks, sidebar, cover art, editlines), and by fixing all those incosistencies (besides the scaling issues with fonts and library controls) this PR is now kind of a GUI round-house kick.
Stay tuned! And please test the effect selector on Windows and MacOS.


@daschuer Please check the library in Shade. I dimmed the table header buttons and IMO all color schemes are much more pleasant now.

Btw: I still need help with the menu for iTunes and Banshee (baseexternallibraryfeature)

**Comment by ronso0** [comment]
Did anyone ever try (or has any hints on how) to style the main menu bar?
This would the icing on the cake...

**Comment by ronso0** [comment]
**Haa!** Got it, I think.
Instead of this complex passing around of m_pSideBarWidget I can parent the menu on the fly to `QApplication::focusWidget()`. will test this, it would simplify the entire story and solve the issue with iTunes and Banshee..

**Comment by ronso0** [comment]
I tested the suggested parent approach `m_pMenu->setParent(anyWidget)` with the Crates feature but this creates more issues than it could potentially avoid:
* menu pops up at wrong place (cursorPos relative to widget 0,0 + global cursorPos)
* menu is cropped by parent widget's geometry
* menu won't close on selection or if I click outside the menu

That said, I'll have to try something else.

**Comment by ronso0** [comment]
I think I've found a way to address the external features, as well.
Nevertheless, I'll split off the skin changes from this PR and leave only a small qss snippet for testing that can easily be removed, because the skin changes are huge and I'd like to include the new AutoDJ drop-down.

**Comment by ronso0** [comment]
Skins styles: #2337 
C++ changes: #2338

**Comment by ronso0** [comment]
Thanks for your help!
